src/hip_module.cpp

```diff
@@ -386,6 +386,118 @@ hipError_t hipExtLaunchMultiKernelMultiDevice(hipLaunchParams* launchParamsList,
     return ihipLogStatus(result);
 }
 
+namespace {
+// kernel for initializing GWS
+// nwm1 is the total number of wavefronts minus 1 and rid is the GWS resource id
+__global__ void init_gws(uint nwm1, uint rid) {
+    __ockl_gws_init(nwm1, rid);
+}
+}
+
+hipError_t hipLaunchCooperativeKernel(const void* f, dim3 gridDim,
+        dim3 blockDimX, void** kernelParams, unsigned int sharedMemBytes,
+        hipStream_t stream) {
+
+    HIP_INIT_API(hipLaunchCooperativeKernel, f, gridDim, blockDimX, kernelParams, sharedMemBytes, stream);
+    hipError_t result;
+
+
+    if ((f == nullptr) || (stream == nullptr) || (kernelParams == nullptr)) {
+        return ihipLogStatus(hipErrorNotInitialized);
+    }
+
+    auto ctx = ihipGetTlsDefaultCtx();
+    if (ctx == nullptr) {
+        return ihipLogStatus(hipErrorInvalidDevice);
+    }
+    int deviceId = ctx->getDevice()->_deviceId;
+    ihipDevice_t* currentDevice = ihipGetDevice(deviceId);
+
+    if (!currentDevice->_props.cooperativeLaunch) {
+        return ihipLogStatus(hipErrorInvalidConfiguration);
+    }
+
+    // Prepare the kernel descriptor for initializing the GWS
+    hipFunction_t gwsKD = hip_impl::get_program_state().kernel_descriptor(
+            reinterpret_cast<std::uintptr_t>(&init_gws),
+            hip_impl::target_agent(stream));
+
+    if (gwsKD == nullptr) {
+        return ihipLogStatus(hipErrorInvalidValue);
+    }
+    hip_impl::kernargs_size_align gwsKargs =
+            hip_impl::get_program_state().get_kernargs_size_align(
+                    reinterpret_cast<std::uintptr_t>(&init_gws));
+
+    gwsKD->_kernarg_layout = *reinterpret_cast<const std::vector<
+            std::pair<std::size_t, std::size_t>>*>(gwsKargs.getHandle());
+
+    // Prepare the kernel descriptor for the main kernel
+    hipFunction_t kd = hip_impl::get_program_state().kernel_descriptor(
+            reinterpret_cast<std::uintptr_t>(f),
+            hip_impl::target_agent(stream));
+    if (kd == nullptr) {
+        return ihipLogStatus(hipErrorInvalidValue);
+    }
+    hip_impl::kernargs_size_align kargs =
+            hip_impl::get_program_state().get_kernargs_size_align(
+                    reinterpret_cast<std::uintptr_t>(f));
+
+    kd->_kernarg_layout = *reinterpret_cast<const std::vector<
+            std::pair<std::size_t, std::size_t>>*>(kargs.getHandle());
+
+
+    void *gwsKernelParams[2];
+    uint nwm1, rid;
+    rid = 0;
+
+    const uint32_t wavefrontSize = 1 << kd->_header->wavefront_size;
+    uint32_t alignedBlockDim = ((blockDimX.x * blockDimX.y * blockDimX.z) + wavefrontSize - 1) &
+                                ~(wavefrontSize - 1);
+
+    // calculate total number of wavefronts minus 1 for the main kernel
+    nwm1 = ((gridDim.x * gridDim.y * gridDim.z) *
```

Comment:
This barrier should be initialized with the number of work groups minus 1 since that is what the sync is expecting.

Reply:
changed the initialization of the barrier to the number of work groups.